### PR TITLE
Lab 02 - Sort algorithm implementation

### PR DESCRIPTION
**Pull request by JackLeeHal**
Fixes #42 

#### Changes proposed in this PR:

* Implement bubble sort
* Implement heap sort
* Add test cases

### CODE REVIEW

File: 02_bubble/jacklee/main.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+func main() {
+	fmt.Fprint(out, bubble([]int{3, 2, 1, 5}))
+}
+
+func bubble(origin []int) []int {
+	result := make([]int, len(origin))
+	copy(result, origin)
+	arrLen := len(origin)
+	for i := 0; i < arrLen-1; i++ {
+		for j := 0; j < arrLen-1-i; j++ {
```
Comment (anzdaddy) [inline 02_bubble/jacklee/main.go:20]: If `i` counts down, you can simplify `j`'s loop. https://github.com/anz-bank/go-training/pull/114#discussion_r261849435

File: 02_bubble/jacklee/main.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+func heapSort(origin []int) []int {
+	heap := make([]int, len(origin))
+	copy(heap, origin)
+	makeHeap(heap)
+	result := make([]int, len(origin))
+	heapLen := len(origin)
+	for i := 0; i < heapLen; i++ {
+		swap(0, len(heap)-1, heap)
+		down(0, len(heap)-1, heap)
+		n := len(heap)
```
Comment (anzdaddy) [inline 02_bubble/jacklee/main.go:38]: Move this to the top of the loop and use it when calling `swap` and `down`.

File: 02_bubble/jacklee/main.go [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+		x := heap[n-1]
+		heap = heap[0 : n-1]
+		result[i] = x
+	}
+	return result
+}
+
+func makeHeap(origin []int) {
+	n := len(origin)
+	for i := n/2 - 1; i >= 0; i-- {
```
Comment (anzdaddy) [inline 02_bubble/jacklee/main.go:48]: It feels like this should be `(n - 1)/2`. I could be wrong, though. I haven't drilled down to check.

File: 02_bubble/jacklee/main.go [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+		if !less(j, i, heap) {
+			break
+		}
+		swap(i, j, heap)
+		i = j
+	}
+}
+
+func less(a, b int, heap []int) bool {
```
Comment (anzdaddy) [inline 02_bubble/jacklee/main.go:73]: I think this complicates things unnecessarily. `!(heap[j1] < heap[j2])` (or `heap[j1] >= heap[j2]`) is more straightforward.

File: 02_bubble/jacklee/main_test.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	//Then
+	expected := "[1 2 3 5]"
+	actual := buf.String()
+	r.Equalf(expected, actual, "Unexpected output")
+}
+
+var testCases = []struct {
+	in  []int
+	out []int
+}{
```
Comment (anzdaddy) [inline 02_bubble/jacklee/main_test.go:28]: This one is largely a matter of personal preference, but in this case I would prefer a one-liner:
```go
var testCases = []struct{in, out []int}{
	…
}
```

**Review by anzdaddy — CHANGES_REQUESTED** [review]

**Comment by juliaogris** [comment]
Sorry, I will close this PR now in preparation for the next training in a few weeks in Melbourne - feel free to re-open it if you want to continue work on it.

**Pull request closed without merging**